### DISCUSSION TITLE
Policies don't work on the server with Supabase client (avatars storage example from quickstart)

### DISCUSSION THREAD

**Discussion by bennettdams** [supabase/supabase#3642] [Questions] [answered]
I would like to use the Supabase client on the server (in my case an API route of Next.js), but the policies don't seem to work.

Is there any example for the policies one would need for using storage for a simple "avatar" use case? There are a lot of templates when creating policies and I think I know how to write them, but I don't think I understand why my policies are not working when I use the Supabase client on the server.

### What I want to do:

There's a great [quickstart for Next.js](https://supabase.io/docs/guides/with-nextjs#bonus-profile-photos), but I can't make the storage work with the policies I think I'd need.

Here's what I want to achieve:

- save pictures (jpg, png) in storage
- allow public read access to all avatar pictures
- allow a user to upload (and change) his avatar

```ts
// this function is used on the Next.js API route (so NodeJS, instead of the browser)
export async function uploadAvatarSupabase(
  filepath: string,
  picture: File
): Promise<void> {
  try {
    const { error } = await supabase.storage
      .from("avatars")
      .upload(`public/${filepath}`, picture, {
        upsert: true,
        contentType: 'image/jpg',
      })
    if (error) throw error
  } catch (error) {
    throw new Error(
      `Error while uploading avatar file: ${error.message || error}`
    )
  }
}
```

My preferred policies for this use case:

- public read access, even for unauthenticated users
- allow upload for authenticated users

On the server, no matter how the user wants to upload the file, I do validation and change the file name to the user ID. That way, I can make sure that one user can only change his own avatar.

### What works for the upload

- Option 1: Set the storage policies as public access (insert, update, delete) - I obviously can't use this
- Option 2: Set the storage policies to allow access for authenticated users (insert, update, delete) (which is how I want to use them in production) and then execute the upload in the browser

### What does NOT work for the upload
Using the right policies (from Option 2) and then, instead of executing the upload from the browser, executing the upload from the server (a Next.js API route, so NodeJS). The error is speaking for itself: `new row violates row-level security policy for table "objects"` - so I'm violating the policies.

**Comment by VictorPeralta** [comment, marked as answer]
Are you using the service key on the server, or the anon/user jwt?

**Reply by bennettdams** [reply]
Mh, interesting. I'm using the anon key, and I thought the service key should [only be used "as an admin"](https://supabase.io/docs/guides/auth/row-level-security#never-use-a-service-key-on-the-client). 
You are right, I could safely use that on the server, but I thought this can also be done as a "user" server-sided - in the end it's an action that was initiated by a user.
I didn't read anything regarding this on the docs, would using the service key server-sided be the way to go?

**Reply by VictorPeralta** [reply]
Oh I understand what you want now. 

Try the solution in this thread, you can act as a user serverside like this:

https://github.com/supabase/supabase/discussions/1094#discussioncomment-714633

**Reply by bennettdams** [reply]
@VictorPeralta That worked, thanks a lot!
